File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    gift_log = {}
    total_gift = {}
    next_month = [0] * len(friends)
                           
    for i in friends:
        temp = {}
        total_gift[i] = 0
        for j in friends:
            temp[j] = 0
        gift_log[i] = temp
    
    for gift in gifts:
        give_freiend, take_friend = gift.split()
        gift_log[give_freiend][take_friend] += 1
        total_gift[give_freiend] += 1
        total_gift[take_friend] -= 1
    
    for i in range(len(friends)):
        for j in range(i, len(friends)):
            if i == j:
                continue
            if gift_log[friends[i]][friends[j]] > gift_log[friends[j]][friends[i]]:
                next_month[i] +=1
            elif gift_log[friends[i]][friends[j]] < gift_log[friends[j]][friends[i]]:
                next_month[j] +=1
            else:
                if total_gift[friends[i]] > total_gift[friends[j]]:
                    next_month[i] +=1
                elif total_gift[friends[i]] < total_gift[friends[j]]:
                    next_month[j] +=1
        
    answer = max(next_month)
    return answer

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;
        int[][] arr = new int[friends.length][friends.length];
        List<String> friendList = Arrays.asList(friends);
        for (int i = 0; i<gifts.length; i++) {
            String[] input = gifts[i].split(" ");
            String A = input[0];
            String B = input[1];
            
            arr[friendList.indexOf(A)][friendList.indexOf(B)]++;
        }
        
        int[] point = new int[friends.length];
        for (int i = 0; i<friends.length; i++) {
             for (int j =0; j<friends.length; j++) {
                 point[i] += arr[i][j];
                 point[i] -= arr[j][i];
             }
        }
        
        for (int i = 0; i<friends.length; i++) {
            int temp = 0;
            for (int j =0; j<friends.length; j++) {
                if (i == j) {
                    continue;
                }
                if (arr[i][j] == arr[j][i]) {
                    if (point[i] > point[j]){
                        temp++;
                    }
                }else if (arr[i][j] > arr[j][i]) {
                    temp++;
                }
            }
            answer = Math.max(answer, temp);
        } 
            
        return answer;
    }
}